### DISCUSSION TITLE
Binding mouse-events to suffixes

### DISCUSSION THREAD

**Discussion by dalanicolai** [magit/transient#160] [Support] [answered]
Hi,

I am sorry to bother you here, but I do not know any better alternative? I am just hoping that maybe you understand this issue and that maybe you can tell me why this does not work. I am asking this again for the [sketch-mode](https://github.com/dalanicolai/sketch-mode) package, which in the meantime is also on [elpa devel](https://elpa.gnu.org/devel/).

Anyway, let me try to keep it short. I am trying to create a command/suffix that contains a `track-mouse`. This functionality would allow me to 'interactively' draw rectangles for selecting objects. The following snippet provides a very simple example:
```
(transient-define-suffix test-mouse ()
  (interactive)
  (track-mouse
    (let ((event (read-event)))
      (while (not (eq (car event) 'S-drag-mouse-1))
        (setq event (print (read-event)))))))


(transient-define-prefix transient-test ()
  "Some Emacs magic"
  :transient-suffix     'transient--do-call
  :transient-non-suffix 'transient--do-stay
  ["Commands"
   [([S-down-mouse-1] "Select" test-mouse)]])
```
Now this snippet seems to work correctly.

However, when I try the same with the pointer on an image, then the event type gets a `sketch` prefix, so that I have to bind it using `[sketch S-down-mouse-1] `.

As you possibly remember, this also work correctly for drag and simple click events, because that is what sketch-mode uses to draw lines and text. But when I try this, over the image, using the `test-mouse` suffix as shown above, then I get the following error message:

`Error in post-command-hook (transient--post-command): (wrong-type-argument keymapp nil)`

I have no idea what transient is doing exactly in the `transient--post-command` function and what kind of keymap it expects.
I hope you can somehow explain it/quickly clear this up to help me find it out, or even better if you already have an idea what could cause this.
Unfortunatey, I can not trigger a backtrace with `toggle-debug-on-error`.

Of course, I do not expect any priority or something by contacting you here; I understand you are busy enough.
I just don't think this counts as a bug, and also this question is not very 'appropriate' for stack-exchange.
Of course, I am happy to move this discussion/question to any of those places.
But I am just hoping that maybe you immediately have got some idea/suspicion what causes this, and provide some suggestion/hints how I could fix/work around it.

I am using magit version: Magit 20210928.1325 [>= 3.2.1], Git 2.31.1, Emacs 28.0.50, gnu/linux
if that is of any relevance.

Anyway, whatever the outcome, thanks a lot!

ps. I see that this discussion area has a `show and tell` section. I think sketch contains some nice examples for how to use transient.
For example, I am setting values of infixes, that are dependent on the values of values of other infixes (the layer infixes).
Also, I think the object and color infixes provide nice examples for developers. So I think it might be handy to add it to the `show and tell` section? What do you think?

**Comment by tarsius** [comment]
This is the right place to bring this up.

> So I think it might be handy to add it to the show and tell section?

That's a great idea. So far I have been suggesting that examples be added to the wiki but maybe show&tell works better, so please give it a try.

**Reply by dalanicolai** [reply]
Okay great! Will do that then. 

In the mean time, I found out that if I just use a keybinding (i.e. not mouse) to bind the `test-mouse` suffix of the example in the comment, then I can track the mouse just fine. So the priority for this question is quite low, but it would be much nicer if the user does not have to press a key first before starting to select some object(s).

**Comment by dalanicolai** [comment]
It looks like the issue with the `Error in post-command-hook (transient--post-command): (wrong-type-argument keymapp nil)` is simply solved by creating the `test-mouse` command with `defun` instead of using `transient-define-suffix`.
@tarsius So you could delete this 'discussion' or leave it here for documentation.

**Reply by tarsius** [reply]
Leave it but please point me to the best version of you project for me to try out, because at this point I only understood "I try to do something with the mouse and it did not work". I'll need more context.

That would be useful for me even if it now works to your satisfaction because I did not consider mouse-events at all when designing transient and if this "just works" (modulo some ugly hacks or not) then that would be interesting to see and maybe even give me ideas on how to improve "support" for mice.

**Reply by dalanicolai** [reply, marked as answer]
Ah well, "annoyingly" I find now that I was wrong. I was trying to create a more complete example, and then I found that the `transient-define-suffix` and `defun` do seem to have the same effect. Both work almost fine in the example below, even when 'dragging` the mouse over an image.
As far as I know I have tried the simple example above of the first comment on an image that I created in a similar way as in the following example:
```
(require 'transient)
(require 'svg)

(defun test-mouse ()
  (interactive)
  (track-mouse
    (let ((event (read-event)))
      (while (not (eq (car event) 'drag-mouse-1))
        (setq event (print (read-event)))))))


(transient-define-prefix transient-test ()
  "Some Emacs magic"
  :transient-suffix     'transient--do-call
  ["Commands"
   [([sketch down-mouse-1] "Select" test-mouse)]])

(let ((width 400)
      (height 400))
  (setq svg (svg-create width height))
  (svg-rectangle svg 0 0 width height :fill "blue")
  (insert-image (svg-image svg
                           :pointer 'hand
                           :map `(((rect . ((0 . 0) . (,(dom-attr svg 'width) . ,(dom-attr svg 'height))))
                                   sketch))))
  (transient-test))
```
but as now I only see a `Error in pre-command-hook (transient--pre-command): (wrong-type-argument (or eieio-object class) nil obj)`
message, I guess I must have done something different last time; maybe I was just using the more elaborate transient of sketch mode (and added this suffix there).
In `sketch-mode` calling the `drawing` suffix (i.e. when drawing with the mouse), still messages a pre- and a post-command hook, but fortunately, these error messages do not obstruct the drawing process (anymore).

Finally, the 'best' version of sketch-mode should be in the `main` branch of the [sketch-mode](https://github.com/dalanicolai/sketch-mode) repo.
Please ignore how the code looks, I am no professional coder, and I am currently just trying to make things work. Anyway, I think the package is already quite useful for quick small sketches...

**Comment by tarsius** [comment]
#160, #162 and #165 are related, and concern https://github.com/dalanicolai/sketch-mode.
